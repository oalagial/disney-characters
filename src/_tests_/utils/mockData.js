export const disneyCharacters = {
  count: 50,
  data: [
    {
      allies: [],
      enemies: [],
      films: [],
      imageUrl:
        'https://static.wikia.nocookie.net/disney/images/6/61/Olu_main.png',
      name: "'Olu Mel",
      parkAttractions: [],
      shortFilms: [],
      tvShows: [],
      url: 'https://api.disneyapi.dev/characters/6',
      videoGames: [],
      _id: 6,
    },
    {
      allies: [],
      enemies: [],
      films: [],
      imageUrl:
        'https://static.wikia.nocookie.net/disney/images/8/80/Profile_-_627.png',
      name: '627',
      parkAttractions: [],
      shortFilms: [],
      tvShows: ['Lilo & Stitch: The Series', 'Stitch!'],
      url: 'https://api.disneyapi.dev/characters/10',
      videoGames: ['Disney Tsum Tsum (game)'],
      _id: 10,
    },
  ],
  nextPage: 'https://api.disneyapi.dev/characters?page=2',
  totalPages: 149,
}
