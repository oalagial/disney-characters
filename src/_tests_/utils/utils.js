import { Provider } from 'react-redux'
import { configureStore } from '@reduxjs/toolkit'
import modalReducer from '../../features/modal/modalSlice'
import { render } from '@testing-library/react'

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { listReducers: modalReducer },
      preloadedState,
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}
