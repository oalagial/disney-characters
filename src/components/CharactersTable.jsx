import React from 'react'
import Table from '@mui/material/Table'
import TableBody from '@mui/material/TableBody'
import TableCell, { tableCellClasses } from '@mui/material/TableCell'
import TableContainer from '@mui/material/TableContainer'
import TableHead from '@mui/material/TableHead'
import TableRow from '@mui/material/TableRow'
import Paper from '@mui/material/Paper'
import { Box, styled, TablePagination, TableSortLabel } from '@mui/material'
import U from '../utils/utils.js'
import { useDispatch } from 'react-redux'
import { openCharacterModal } from '../features/modal/modalSlice'
import { visuallyHidden } from '@mui/utils'

const StyledTableCell = styled(TableCell)(({ theme }) => ({
  [`&.${tableCellClasses.head}`]: {
    backgroundColor: 'dimgray',
    color: 'gold',
    fontSize: 20,
  },
  [`&.${tableCellClasses.body}`]: {
    fontSize: 14,
  },
}))

const StyledTableRow = styled(TableRow)(({ theme }) => ({
  '&:nth-of-type(odd)': {
    backgroundColor: theme.palette.action.hover,
  },
  // hide last border
  '&:last-child td, &:last-child th': {
    border: 0,
  },
}))

const CharactersTable = ({
  disneyCharacters,
  page,
  pageSize,
  handlePageChanged,
  handlePageSizeChanged,
}) => {
  const [order, setOrder] = React.useState('asc')
  const [orderBy, setOrderBy] = React.useState()
  const dispatch = useDispatch()

  const handleRequestSort = (event, property) => {
    const isAsc = orderBy === property && order === 'asc'
    setOrder(isAsc ? 'desc' : 'asc')
    setOrderBy(property)
  }

  const createSortHandler = (property) => (event) => {
    handleRequestSort(event, property)
  }

  const totalPages = disneyCharacters?.totalPages || 0

  return (
    <Paper sx={{ width: '100%', mb: 2 }}>
      <TableContainer component={Paper}>
        <Table sx={{ minWidth: 650 }} aria-label="simple table">
          <TableHead>
            <TableRow>
              <StyledTableCell
                key="name"
                align="center"
                sortDirection={orderBy === 'name' ? order : false}
              >
                <TableSortLabel
                  active={orderBy === 'name'}
                  direction={orderBy === 'name' ? order : 'asc'}
                  onClick={createSortHandler('name')}
                >
                  Name
                  {orderBy === 'name' ? (
                    <Box component="span" sx={visuallyHidden}>
                      {order === 'desc'
                        ? 'sorted descending'
                        : 'sorted ascending'}
                    </Box>
                  ) : null}
                </TableSortLabel>
              </StyledTableCell>
              <StyledTableCell align="center">
                Number of TV Shows
              </StyledTableCell>
              <StyledTableCell align="center">
                Number of Video Games
              </StyledTableCell>
              <StyledTableCell align="center">Allies</StyledTableCell>
              <StyledTableCell align="center">Enemies</StyledTableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {U.stableSort(
              disneyCharacters.data,
              U.getComparator(order, orderBy)
            )
              // .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
              .map((row, index) => {
                return (
                  <StyledTableRow
                    key={row.name + row._id}
                    data-testid={row.name + row._id}
                    style={{ cursor: 'pointer' }}
                    onClick={() =>
                      dispatch(
                        openCharacterModal({
                          id: row._id,
                          name: row.name,
                          imageUrl: row.imageUrl,
                          tvShows: row.tvShows,
                          videoGames: row.videoGames,
                        })
                      )
                    }
                  >
                    <TableCell align="center">{row.name}</TableCell>
                    <TableCell align="center">{row.tvShows.length}</TableCell>
                    <TableCell align="center">
                      {row.videoGames.length}
                    </TableCell>
                    <TableCell align="center">
                      {row.allies.length > 0
                        ? row.allies.map((ally) => (
                            <TableRow
                              key={ally}
                              sx={{
                                '&:last-child td, &:last-child th': {
                                  border: 0,
                                },
                              }}
                            >
                              <TableCell component="th" scope="row">
                                {ally}
                              </TableCell>
                            </TableRow>
                          ))
                        : '-'}
                    </TableCell>
                    <TableCell align="center">
                      {row.enemies.length > 0
                        ? row.enemies.map((enemy) => (
                            <TableRow
                              key={enemy}
                              sx={{
                                '&:last-child td, &:last-child th': {
                                  border: 0,
                                },
                              }}
                            >
                              <TableCell component="th" scope="row">
                                {enemy}
                              </TableCell>
                            </TableRow>
                          ))
                        : '-'}
                    </TableCell>
                  </StyledTableRow>
                )
              })}
          </TableBody>
        </Table>
      </TableContainer>
      <TablePagination
        rowsPerPageOptions={[10, 20, 50, 100, 200, 500]}
        count={pageSize * totalPages}
        page={page - 1}
        rowsPerPage={pageSize}
        onPageChange={(e, page) => handlePageChanged(page + 1)}
        onRowsPerPageChange={(e) => {
          handlePageChanged(1)
          handlePageSizeChanged(e.target.value)
        }}
      />
    </Paper>
  )
}

export default CharactersTable
