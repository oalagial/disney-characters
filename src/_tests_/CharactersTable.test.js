import React from 'react'
import { render, screen, within } from '@testing-library/react'
import { Provider } from 'react-redux'
import { store } from '../app/store'
import renderer from 'react-test-renderer'
import CharactersTable from '../components/CharactersTable'
import { disneyCharacters } from './utils/mockData'

describe('CharactersTable', () => {
  test('renders correctly', async () => {
    const { getByText, getByRole } = render(
      <Provider store={store}>
        <CharactersTable disneyCharacters={disneyCharacters} />
      </Provider>
    )

    const tableElement = screen.getByRole('table')
    expect(tableElement).toBeInTheDocument()

    const columnHeader = screen.getAllByRole('columnheader')
    expect(columnHeader[0].textContent).toBe('Name')
    expect(columnHeader[1].textContent).toBe('Number of TV Shows')
    expect(columnHeader[2].textContent).toBe('Number of Video Games')
    expect(columnHeader[3].textContent).toBe('Allies')
    expect(columnHeader[4].textContent).toBe('Enemies')
  })

  test('renders correctly all cells', async () => {
    const { getByText, getByRole, getAllByRole } = render(
      <Provider store={store}>
        <CharactersTable disneyCharacters={disneyCharacters} />
      </Provider>
    )

    const cells = screen.getAllByRole('cell')
    expect(cells[0].textContent).toBe(disneyCharacters.data[0].name)
    expect(cells[1].textContent).toBe(
      disneyCharacters.data[0].tvShows.length.toString()
    )
    expect(cells[2].textContent).toBe(
      disneyCharacters.data[0].videoGames.length.toString()
    )
  })

  test('Matches DOM Snapshot', () => {
    const domTree = renderer
      .create(
        <Provider store={store}>
          <CharactersTable
            disneyCharacters={disneyCharacters}
            page={1}
            pageSize={50}
          />
        </Provider>
      )
      .toJSON()
    expect(domTree).toMatchSnapshot()
  })
})
