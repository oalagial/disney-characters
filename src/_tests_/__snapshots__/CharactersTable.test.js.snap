// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharactersTable Matches DOM Snapshot 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1mxz8qt-MuiPaper-root"
>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
  >
    <table
      aria-label="simple table"
      className="MuiTable-root css-quj9j6-MuiTable-root"
      role={null}
    >
      <thead
        className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1e7p7ua-MuiTableCell-root"
            scope="col"
          >
            <span
              className="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role="button"
              tabIndex={0}
            >
              Name
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                data-testid="ArrowDownwardIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1e7p7ua-MuiTableCell-root"
            scope="col"
          >
            Number of TV Shows
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1e7p7ua-MuiTableCell-root"
            scope="col"
          >
            Number of Video Games
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1e7p7ua-MuiTableCell-root"
            scope="col"
          >
            Allies
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1e7p7ua-MuiTableCell-root"
            scope="col"
          >
            Enemies
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root css-1enotpu-MuiTableRow-root"
          data-testid="'Olu Mel6"
          onClick={[Function]}
          role={null}
          style={
            Object {
              "cursor": "pointer",
            }
          }
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
          >
            'Olu Mel
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
          >
            0
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
          >
            0
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
          >
            -
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
          >
            -
          </td>
        </tr>
        <tr
          className="MuiTableRow-root css-1enotpu-MuiTableRow-root"
          data-testid="62710"
          onClick={[Function]}
          role={null}
          style={
            Object {
              "cursor": "pointer",
            }
          }
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
          >
            627
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
          >
            2
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
          >
            1
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
          >
            -
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
          >
            -
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <td
    aria-sort={null}
    className="MuiTableCell-root MuiTableCell-sizeMedium MuiTablePagination-root css-9wcvca-MuiTableCell-root-MuiTablePagination-root"
    colSpan={1000}
  >
    <div
      className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-78c6dr-MuiToolbar-root-MuiTablePagination-toolbar"
    >
      <div
        className="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
      />
      <p
        className="MuiTablePagination-selectLabel css-pdct74-MuiTablePagination-selectLabel"
        id=":r1:"
      >
        Rows per page:
      </p>
      <div
        className="MuiInputBase-root MuiInputBase-colorPrimary css-16c50h-MuiInputBase-root-MuiTablePagination-select"
        onClick={[Function]}
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby=":r1: :r0:"
          className="MuiSelect-select MuiTablePagination-select MuiSelect-standard MuiInputBase-input css-194a1fa-MuiSelect-select-MuiInputBase-input"
          id=":r0:"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          50
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          onAnimationStart={[Function]}
          onChange={[Function]}
          tabIndex={-1}
          value={50}
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiTablePagination-selectIcon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <p
        className="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
      >
        1â€“50 of 7450
      </p>
      <div
        className="MuiTablePagination-actions"
      >
        <button
          aria-label="Go to previous page"
          className="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Go to previous page"
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="KeyboardArrowLeftIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
            />
          </svg>
        </button>
        <button
          aria-label="Go to next page"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          title="Go to next page"
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="KeyboardArrowRightIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
            />
          </svg>
        </button>
      </div>
    </div>
  </td>
</div>
`;
